src/Data/Bin/Ordering.agda

```diff
@@ -0,0 +1,791 @@
+
+≡⇒≮ :  _≡_ ⇒ _≮_
+≡⇒≮ x≡y x<y =  <⇒≢ x<y x≡y
+
+≡⇒≯ :  _≡_ ⇒ _≯_
+≡⇒≯ x≡y x>y =  >⇒≢ x>y x≡y
+
+<-trans : Transitive _<_
+<-trans {_}      {_}      {zero}     _  ()
+<-trans {zero}   {_}      {2[1+ _ ]} _  _  =  0<even
+<-trans {_}      {zero}   {_}        ()
+<-trans {zero}   {_}      {1+[2 _ ]} _  _  =  0<odd
+
+<-trans {2[1+ x ]} {2[1+ y ]} {2[1+ z ]} (even<even x<y) (even<even y<z) =
+                                                    even<even (<-trans x<y y<z)
+<-trans {2[1+ x ]} {2[1+ y ]} {1+[2 z ]} (even<even x<y) (even<odd y<z) =
+                                                    even<odd (<-trans x<y y<z)
+<-trans {2[1+ x ]} {1+[2 y ]} {2[1+ z ]} (even<odd x<y) (odd<even (inj₁ y<z)) =
+                                                    even<even (<-trans x<y y<z)
+<-trans {2[1+ x ]} {1+[2 y ]} {2[1+ .y ]} (even<odd x<y)
+                                          (odd<even (inj₂ refl)) =  even<even x<y
+<-trans {2[1+ x ]} {1+[2 y ]} {1+[2 z ]} (even<odd x<y) (odd<odd y<z) =
+                                                        even<odd (<-trans x<y y<z)
+<-trans {1+[2 x ]} {2[1+ y ]} {2[1+ z ]} (odd<even (inj₁ x<y)) (even<even y<z) =
+                                                   odd<even (inj₁ (<-trans x<y y<z))
+<-trans {1+[2 x ]} {2[1+ .x ]} {2[1+ z ]} (odd<even (inj₂ refl))
+                                          (even<even x<z) =  odd<even (inj₁ x<z)
+<-trans {1+[2 x ]} {2[1+ y ]} {1+[2 z ]} (odd<even (inj₁ x<y)) (even<odd y<z) =
+                                                         odd<odd (<-trans x<y y<z)
+<-trans {1+[2 x ]} {2[1+ .x ]} {1+[2 z ]} (odd<even (inj₂ refl)) (even<odd x<z) =
+                                                                 odd<odd x<z
+<-trans {1+[2 x ]} {1+[2 y ]} {2[1+ z ]} (odd<odd x<y) (odd<even (inj₁ y<z)) =
+                                                  odd<even (inj₁ (<-trans x<y y<z))
+<-trans {1+[2 x ]} {1+[2 y ]} {2[1+ .y ]} (odd<odd x<y) (odd<even (inj₂ refl)) =
+                                                             odd<even (inj₁ x<y)
+<-trans {1+[2 x ]} {1+[2 y ]} {1+[2 z ]} (odd<odd x<y) (odd<odd y<z) =
+                                                       odd<odd (<-trans x<y y<z)
+
+<⇒≯ : _<_ ⇒ _≯_
+<⇒≯ {x} {y} x<y y<x =  <-irrefl refl (<-trans {x} {y} {x} x<y y<x)
+
+>⇒≮ : _>_ ⇒ _≮_
+>⇒≮ {x} {y} y<x =  <⇒≯ {y} {x} y<x
+
+<⇒≤ : _<_ ⇒ _≤_
+<⇒≤ = inj₁
+
+------------------------------------------------------------------------------
+-- resolving inequality comparison
+------------------------------------------------------------------------------
+
+open Tri
+
+<-cmp :  ∀ (x y) → Tri (x < y) (x ≡ y) (x > y)
+<-cmp zero     zero      = tri≈ ≮0     refl  ≮0
+<-cmp zero     2[1+ _ ]  = tri< 0<even (λ()) ≮0
+<-cmp zero     1+[2 _ ]  = tri< 0<odd  (λ()) ≮0
+<-cmp 2[1+ _ ] zero      = tri> (λ())  (λ()) 0<even
+<-cmp 2[1+ x ] 2[1+ y ]  with <-cmp x y
+... | tri< x<y _   _ =  tri< lt (<⇒≢ lt) (<⇒≯ lt)  where lt = even<even x<y
```

Comment:
Why use a `where` here?

Reply:
In order to avoid the three copies of the string "(even<even x<y)" in the right hand side.
Really, do you like to read it as repeated three times? (four times?)
Also the type checker (or compiler) may need, to discover that the term
```
  tri< (even<even x<y) (<⇒≢ (even<even x<y)) (<⇒≯ (even<even x<y))
```
is an instance of ``tri< Z (<⇒≢ Z) (<⇒≯ Z)``.
And it is easier to do this for the term
``tri< lt (<⇒≢ lt) (<⇒≯ lt)  where lt = even<even x<y``.

Reply:
That's a good reason, hadn't spotted that, apologies!